src/Foundatio.AzureServiceBus/Utility/AuthHelper.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Threading.Tasks;
+using Foundatio.Logging;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+namespace Foundatio.AzureServiceBus.Utility
+{
+    public class TokenModel {
+        public string TokenValue { get; set; }
+        public DateTime TokenExpiresAtUtc { get; set; }
+    }
+
+
+    public static class AuthHelper {
+        public static async Task<TokenModel> GetToken(string tokenValue, DateTime tokenExpiresAtUtc, string tenantId, string clientId, string clientSecret) {
+            try {
+
+                if (String.IsNullOrEmpty(tenantId) || String.IsNullOrEmpty(clientId) || String.IsNullOrEmpty(clientSecret)) {
+                    return null;
+                }
+
+                var tokenModel = new TokenModel() {
+                    TokenValue = tokenValue,
+                    TokenExpiresAtUtc = tokenExpiresAtUtc
+                };
+                // Check to see if the token has expired before requesting one.
+                // We will go ahead and request a new one if we are within 2 minutes of the token expiring.
+                if (tokenExpiresAtUtc < DateTime.UtcNow.AddMinutes(-2) || tokenValue == String.Empty) {
+                    Console.WriteLine("Renewing token...");
+
+                    var context = new AuthenticationContext($"https://login.windows.net/{tenantId}");
+
+                    var result = await context.AcquireTokenAsync(
+                        "https://management.core.windows.net/",
+                        new ClientCredential(clientId, clientSecret)
+                    );
+
+                    // If the token isn't a valid string, throw an error.
+                    if (String.IsNullOrEmpty(result.AccessToken)) {
```

Comment:
> String.IsNullOrEmpty

should be 

**s**tring.IsNullOrEmpty

Reply:
@SeanFeldman Thanks Sean for reviewing this. I know I am asking for too much but can you please also take a glance at the functionality on how I am calling AbandonAsync, CompleteAsync using 2 different objects ( IQueueclient and IMessageReceiver )

 await _queueClient.CompleteAsync(entry.Id).AnyContext();
await _messageReceiver.CompleteAsync(entry.Id).AnyContext();

As you noticed this library has 2 ways of pulling the items from the queue:
1) Push Strategy - Which IQueueClient provides by registering RegisterMessageHandler.
2) Pull Strategy - Which IMessageReceive provides using ReceiveAsync.

For some reason, in the new azure service bus dot net core library ReceiveAsync method has been removed from the IQueueClient and that leaves me to use the IMessageReceiver for fetching the item from the queue for the 2nd strategy.

Even though the endpoint for the queue is same if I try to call AbandonAsyc using the IMessageReceiver object for the queue item that has been dequeued using IQueueClient, I get "MessageLockLostException". Similary, If i try to call AbandonAsync using the IQueueClient object for the queue item that has been dequeued using IMessageReceiver, then also i get the same exception.
Because of this reason, I have to maintain some metadata in the QueueEntry to know if the queue item was dequeued using push or pull strategy so that I can call the AbandonAsync or CompleteAsync and even RenewLockAsync using appropriate objects ( either IQueueClient or IMessageReceiver ).

Do you know by any chance why there is a difference in behavior ? Shouldnt there be no exception if i try to call any of the above methods with any object.

Reply:
The older version of the library used SBMP by default. The new one is using AMQP. With SBMP you could receive a message with one receiver and complete with another where with AMQP there's a receiver affinity. Meaning you **have** to complete/abandon/renew your message using the same receiver used to get the message to begin with.

As for the MessageAPI (registering message handler) and `ReceiveAsync`, it's all there, just slightly reshuffled. `ReceiveAsync` is now on `MessageReceiver` as it's considered a more "advanced" option (you control number of messages, write your own message pump, etc) and `QueueClient` is more of a "simple" scenario, intended to be used with the OOTB message pump registered with `RegisterSessionHandler`.

Reply:
`AMQP there's a receiver affinity.` Is that for AMQP in general?

Reply:
For the ASB .NET Standard client. Don't know if that's a general AMQP requirement or not.